### PR TITLE
API for IAB dashboards

### PR DESCRIPTION
**Pull request by dsnyder-fairway**

### CODE REVIEW

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java [hunk 1]
```diff
@@ -23,4 +24,9 @@
         return service.getGroups(user);
     }
 
+    @GetMapping("/{groupId}")
+    public Group getGroup(@AuthenticationPrincipal final User user, @PathVariable final int groupId) {
```
Comment (dsnyder-fairway) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java]: will be used with a `forkJoin` to get the total number of students in the group

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/jdbc/AssessmentItems.java [hunk 2]
```diff
@@ -1,4 +1,6 @@
-package org.opentestsystem.rdw.reporting.common.model;
+package org.opentestsystem.rdw.reporting.common.jdbc;
```
Comment (dsnyder-fairway) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/jdbc/AssessmentItems.java]: moved some of these to their appropriate package. From org.opentestsystem.rdw.reporting.common.model -> org.opentestsystem.rdw.reporting.common.jdbc

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java (hunk 1 above)
Comment (jtreuting) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java]: we don't already have the groups on the client side?

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/group/GroupAssessmentController.java [hunk 3]
```diff
@@ -38,4 +39,12 @@ public GroupAssessment getLatestGroupAssessment(
         return service.getAllAssessments(user, search);
     }
 
+    @GetMapping("/groups/{groupId}/measuredassessments")
+    public List<MeasuredAssessment> getMeasuredAssessments(
+            @AuthenticationPrincipal final User user,
+            final GroupAssessmentSearch search) {
```
Comment (jtreuting) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/group/GroupAssessmentController.java]: is this search mapping from the querystring?  also, the feature does not need the ability for advanced filters, does the API and the SQL support that right now?

Comment (dsnyder-fairway) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/group/GroupAssessmentController.java]: the SQL only needs groupId and schoolYear, but that is all the GroupAssessmentSearch contains

Comment (jtreuting) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/group/GroupAssessmentController.java]: okay cool.  thanks for clarifying

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java (hunk 1 above)
Comment (dsnyder-fairway) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupController.java]: we do the call above to get the list of groups on the main page. I suppose we could use the student count from there instead of this endpoint when we want the the number of students in the group.

File: reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java [hunk 4]
```diff
@@ -37,19 +39,30 @@
...
     @Test
     public void findAllShouldFindAll() {
         assertThat(repository.findAll(userBuilder.permissionsById(groupStatewide).build(), search().build()))
                 .usingRecursiveFieldByFieldElementComparator()
                 .containsOnlyElementsOf(TestData.STUDENT10_GROUP10_ASSESSMENTS_WITH_CUT_POINTS);
+
+        final List<MeasuredAssessment> measuredAssessments = repository.findAllMeasuredAssessments(userBuilder.permissionsById(groupStatewide).build(), search().build());
+        assertThat(measuredAssessments)
+                .usingRecursiveFieldByFieldElementComparator()
+                .containsOnlyElementsOf(TestData.STUDENT10_MEASURED_ASSESSMENTS);
```
Comment (agorina) [inline reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: I expected more test cases for this call.  Do you feel that it covers all possible use cases?

Comment (dsnyder-fairway) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: Maybe.. This test executes the same query as the `findAll` above it. That `findAll` contains an IAB, ICA, and Summative assessment, but this one contains only the IAB. The other test verifies no assessments are found. Am I missing a test to verify a test with more than one level count? What other tests should there be?

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/group/GroupAssessmentRepository.java [hunk 5]
```diff
@@ -1,6 +1,23 @@
...
+    /**
+     * Finds all IAB measured assessments for the given group, assessment grade, school year and permissions
+     *
+     * @param permissionSource the security information to prepare the SQL query parameters for
+     * @param search           the search parameters to limit the results by
+     * @return all IAB measured assessments for the given group, assessment grade, school year and permissions
+     * If the search includes a student group id and {@link PermissionSource} does not have access to the group,
+     * it returns an empty list.
+     */
+    List<MeasuredAssessment> findAllMeasuredAssessments(@NotNull final PermissionSource permissionSource, @NotNull final GroupAssessmentSearch search);
```
[the remark this answers is not in the document]
Comment (esotrope) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/exam/group/GroupAssessmentRepository.java]: should that limit be imposed in the query for now?
then later our assessment/subject service can inform the UI that "measuredAssessmentTypes" is "iab" only or something

File: reporting-service/src/main/resources/application-no-permission.sql.yml [hunk 6]
```diff
@@ -10,12 +10,32 @@ sql:
         sg.name,
         sc.name as school_name,
         sc.id as school_id,
-        s.code as subject_code
+        s.code as subject_code,
+        count(sgm.student_id) as students
```
Comment (agorina) [inline reporting-service/src/main/resources/application-no-permission.sql.yml]: Nit: it should be `count(*)`, here and below.

And I wonder, why this change? Do we need to return counts for all user groups somewhere?

File: reporting-service/src/main/resources/application-no-permission.sql.yml [hunk 7]
```diff
@@ -10,12 +10,32 @@ sql:
...
       from user_student_group usg
         join student_group sg on sg.id = usg.student_group_id
         join school sc on sc.id = sg.school_id
+        join student_group_membership sgm on sgm.student_group_id = usg.student_group_id
```
Comment (esotrope) [inline reporting-service/src/main/resources/application-no-permission.sql.yml]: so if i follow this right - we get the count for the groups on the homepage and just dont display it. we cache the groups with counts and then we look it up again on the individual group page? i'm leaning towards the individual group page getting all the groups and using the counts from there - that page will need to show a select box with all groups anyway right?

File: reporting-service/src/main/resources/application-no-permission.sql.yml (hunk 6 above)
Comment (dsnyder-fairway) [reply reporting-service/src/main/resources/application-no-permission.sql.yml]: yeah this was a little unclear to me the best way to go about this. We could either do this when we get all the groups and just hold onto the counts in cache, or we can do this on every group lookup like `findOneById`. Thoughts? @esotrope @jtreuting conversation will follow here for student count on groups

File: reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java (hunk 4 above)
Comment (agorina) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: What about verifying things like:
1. If a student took an asmt twice - it takes the latests one within the school year
2. It returns the right completed_at
3. I just noticed that we do not filter unscored exams. I missed it in the SQL. I am not sure how we want to handle it, but we need to consider that.  Let me know if you need help with adjusting the SQL. 
4. and more ... 

There are several conditional paths in that SQL and the idea for the repository IT is to test all the cases. I am sure my list above is not complete.

File: reporting-service/src/main/resources/application-no-permission.sql.yml (hunk 6 above)
Comment (agorina) [reply reporting-service/src/main/resources/application-no-permission.sql.yml]: My two cents: if we do not display the counts - do not count in MySQL.

File: reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java (hunk 4 above)
Comment (dsnyder-fairway) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: Thanks @agorina I'll close this PR and work on these

File: reporting-service/src/main/resources/application-no-permission.sql.yml (hunk 6 above)
Comment (dsnyder-fairway) [reply reporting-service/src/main/resources/application-no-permission.sql.yml]: but we will need the group size to display on the IAB pages

File: reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java (hunk 4 above)
Comment (agorina) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: No need to close. You could add the tests later. Up to you.

File: reporting-service/src/main/resources/application-no-permission.sql.yml (hunk 6 above)
Comment (agorina) [reply reporting-service/src/main/resources/application-no-permission.sql.yml]: One group or all groups?

Comment (dsnyder-fairway) [reply reporting-service/src/main/resources/application-no-permission.sql.yml]: Either X times for a using going through X number of groups (`findOneById`) or once (`findAllForUsername`). But I did both here 😆

Comment (jtreuting) [reply reporting-service/src/main/resources/application-no-permission.sql.yml]: it feels to me that we are getting all the groups to populate that dropdown.  so it seems logical to do it at that point and have the counts already.  BUT to Alla's point for performance and load testing reasons since we don't know how often someone will go to the dashboard page and it's not the default maybe we should just call the findOneById and get the count there so it's only calculated when we actually need it.

that is a long rambling way of suggesting the `findOneById` approach and not calculating it for all groups.  let's take the hit when we know we need to.  

P.S @agorina is scared of counting things apparently :)

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+    }
+
+    public static class Builder {
+        private Assessment assessment;
+        private Measures measures;
+        private Integer studentsTested;
+        private Instant completedAt;
+
+        public Builder() {
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java]: shouldn't expose the constructor

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
+package org.opentestsystem.rdw.reporting.common.model;
+
+import java.time.Instant;
+
+public class MeasuredAssessment {
```
Comment (malaffoon) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java]: javadocs, what's a measured assessment?

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java [hunk 10]
```diff
@@ -15,4 +15,12 @@
...
     List<Group> findAllForUsername(String username);
 
+    /**
+     * Find the {@link Group} by group id for the given user name
+     *
+     * @param username the user name
+     * @param groupId the group id
+     * @return a {@link Group}
+     */
+    Group findOneById(String username, int groupId);
```
Comment (malaffoon) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java]: Bit confused by this: doesn't the groupId fully constrain a group? So this will return null if the user isn't a member of the group? Something feels wrong about that. If we do leave it this way then the method name should probably be findOneByUsernameAndId.

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java [hunk 11]
```diff
@@ -15,4 +15,12 @@
      */
     List<Group> findAllForUsername(String username);
 
+    /**
+     * Find the {@link Group} by group id for the given user name
+     *
+     * @param username the user name
+     * @param groupId the group id
+     * @return a {@link Group}
```
Comment (malaffoon) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java]: null if not found?

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupService.java [hunk 12]
```diff
@@ -18,4 +18,12 @@
      */
     Set<Group> getGroups(@NotNull User user);
 
+    /**
+     * Gets the student group by group id that the user has access to
+     *
+     * @param user the {@link User}
+     * @param groupId the group id
+     * @return the group by group id the user has access to
```
Comment (malaffoon) [inline reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupService.java]: throws if not found?

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java (hunk 8 above)
Comment (dsnyder-fairway) [reply common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java]: oops nice catch

File: reporting-service/src/main/resources/application.sql.yml [hunk 13]
```diff
@@ -99,6 +99,55 @@ sql:
           AND (sg.subject_id is null OR a.subject_id = sg.subject_id)
           AND ${sql.reporting.snippet.studentExamPermissionWithEmbargo}
 
+    measuredassessment:
```
Comment (malaffoon) [inline reporting-service/src/main/resources/application.sql.yml]: nit: `measuredAssessment`?

Comment (dsnyder-fairway) [reply reporting-service/src/main/resources/application.sql.yml]: done

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupService.java (hunk 12 above)
Comment (dsnyder-fairway) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupService.java]: yup

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java (hunk 11 above)
Comment (dsnyder-fairway) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java]: yup

File: reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java (hunk 10 above)
Comment (dsnyder-fairway) [reply reporting-service/src/main/java/org/opentestsystem/rdw/reporting/group/GroupRepository.java]: renamed

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java (hunk 9 above)
Comment (dsnyder-fairway) [reply common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java]: added

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java (hunk 8 above)
Comment (dsnyder-fairway) [reply common/src/main/java/org/opentestsystem/rdw/reporting/common/model/MeasuredAssessment.java]: made private

File: reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java (hunk 4 above)
Comment (agorina) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: @jtreuting , could you please clarify item 3? 
Should we exclude exams that have null scores from the query?

As for the rest, we could look at it together on Monday if you still have issues with it.

Comment (jtreuting) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: null scores should not be included.

Comment (agorina) [reply reporting-service/src/test/java/org/opentestsystem/rdw/reporting/exam/group/JdbcGroupAssessmentRepositoryIT.java]: Hmmmm...this would probably require a change to the index too. Will work on it.

**Comment by dsnyder-fairway** [comment]
I'll re-open this once the issues have been resolved
